vanilla_bgv/params.py

```diff
@@ -0,0 +1,19 @@
+"""Parameter containers."""
+
+from dataclasses import dataclass
+from functools import cached_property
+import math
+from vanilla_bgv import number_theory
+
+
+@dataclass(frozen=True)
+class BGVParams:
+  """Parameters for the BGV scheme."""
+
+  poly_mod_degree: int  # referred to as N, a power of 2
+  plaintext_coeff_modulus: int  # referred to as t, a prime
+
+  @cached_property
+  def ntt_modulus(self):
```

Comment:
(I'm assuming this isn't intended to be an actual secure ciphertext space modulus, since you're just choosing the minimal for an NTT, might want to drop a comment that this isn't a secure ciphertext modulus param)

Reply:
Really, I'm confused about this. I understand that BGV sets up the chain of moduli Q = q_0 * ... * q_L, with each equivalent to 1 mod 2N, but this is in the ciphertext space, and it's not clear to me how those moduli relate to the encoding procedure and plaintext space. I should call this `plaintext_ntt_modulus`.

Re-reading it now, it's clear to me that I should be using `2*N` instead of `N` for the poly_mod_degree.

This is how OpenFHE does it, but I'm not quite clear on what `m_initRoot` is, and they are doing the double-CRT representation so there's the extra layer of CRT there...

Reply:
I _think_ that for the purposes of the NTT, the selection of this modulus is simply to ensure you have the roots of unity for the purposes of the transform, and then it is not used after that.

Reply:
> I think that for the purposes of the NTT, the selection of this modulus is simply to ensure you have the roots of unity for the purposes of the transform, and then it is not used after that.

Right, we need that q = 1 mod 2N modulus so that \phi_N(x) mod q can be factored into the product of degree 1 polynomials so you can have a single value slot. 

Yeah, i didn't realize a lot of implementations just choose the smallest such modulus larger than the plaintext rather than the actual ciphertext modulus to make encode/decode easier, which makes sense.

>  it's not clear to me how those moduli relate to the encoding procedure and plaintext space

I think the relationship after that is an emedding since the moduli have the same congruence mod the plaintext modulus

>  I should call this plaintext_ntt_modulus.

yeah! or something like "encoding_ntt_modulus" since it's really only used for that computation

Reply:
I was running into some issues with this that I wasn't able to figure out, so I ended up setting the plaintext modulus and the NTT modulus for encoding to be the same.